Zharkovsky Nikita (2k61g lab1 task8)

Zharkovsky opened the pull request:
№8. Из trie-дерева удалить все слова, которые содержат указанную подстроку.

Zharkovsky committed:
Zharkovsky Nikita Lab1 task8

Zharkovsky committed:
Update Zharkovsky_lab1_task8.cpp

tapoton commented on Zharkovsky Nikita/Zharkovsky_lab1_task8.cpp:
```diff
@@ -0,0 +1,239 @@
+using namespace std;
+
+struct trie {
+	char *value;
```
Для чего вам `value` типа `char*`? По ходу просмотра кода я так и не увидел необходимости в дополнительном слове

Zharkovsky replied:
Параметр value добавлен чисто для демонстрации возможности добавления дополнительных полей для структуры данных, а также для расширения возможностей использования данной структуры.
Например, если мы хотим составить trie-дерево названий продуктов и ассоциировать их с ценами, то данный пункт в структуре будет очень кстати.

tapoton commented on Zharkovsky Nikita/Zharkovsky_lab1_task8.cpp:
```diff
@@ -0,0 +1,239 @@
+	}
+}
+
+void del_substring(struct trie *root, char *word) 
```
Функция не отрабатывает верно. Пример:
```
q
 w
  e
   r
    t
     y (q)
```

вызываем функцию с параметром "qwer", остается дерево:
```
q
 w
  e
```

Zharkovsky replied:
В моем представлении функция должна работать таким образом: 
Создается так называемое дерево. Далее по условию нам нужно удалить все слова с каким-то определенным префиксом. 
Какие слова содержат данный префикс? - те, что включают этот префикс, а если быть еще более точным, то это все слова, составленные в дереве, которые начинаются с последнего символа найденного префикса. 
Т.е например, qwerty и qwerta, мы хотим удалить слова которые включают "we". Следовательно мы находим все слова, которые включают данный префикс(в данном случае оба слова), а далее если мы удалим все слова, которые идут, начиная с последнего символа этого префикса, то в результате окажется, что после удаления не будет слов с данным префиксом, что и требовалось.
Возможно посчитать не правильным, что осталась ветка до конца префикса, НО ее удаление будет зависеть от конкретных условий, потребностей программы. Если нам нужно максимально сохранить дерево, в надежде что в последующем мы еще многократно дополним эту ветку, то есть смысл оставить ее. Если же у нас задача экономии максимально памяти, то да, тогда бы стоило удалить все до узла, который указывает на слово (выше найденного префикса).
Но т.к в условии не было указано подобных условий по этому поводу, решил остановиться на условии максимального сохранения дерева.
И на счет предыдущего вопроса. Мой ответ несколько не корректный. В данной реализации value  символизирует само слово. Т.е если у узла есть value, то это слово, внесенное в дерево, иначе - это промежуточное звено.
Исходя из этого, при последующей работе с программой и структурой, если, например, потребуется вывести все слова, то оставленные пустые ветки(без value) не сыграют никакой роли, т.к слова символизируются не пустым value

tapoton replied:
Исходя из определения trie-дерева, в нем не может быть листа, который завершает некоторый добавленный в дерево ключ.
Как вы сами сказали, наличие value указывает на конец ключа. В приведенном мной примере, value в листе нет.
Соответственно, сам факт того, что вы работаете с trie-деревом, создает для вас требование удалять рудиментные узлы.

Zharkovsky committed:
Update Zharkovsky_lab1_task8.cpp

value символизирует слово. А сама структура теперь удаляет рудиментные элементы.

tapoton commented:
Добавил два ключа "qwerty" и "asdfg". 
Запросил удаление по подстроке "er".
Вывожу дерево: пусто.

Zharkovsky committed:
New version of lab

Zharkovsky committed:
Update Lab1.cpp

tapoton commented:
5. Задан набор записей следующей структуры: номер телефона, ФИО, адрес. По номеру телефона найти сведения о ФИО владельца и его адресе.
